File: app/src/main/java/com/example/myanimedata/ui/FavoriteCharacterFragment.java
package com.example.myanimedata.ui;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myanimedata.R;
import com.example.myanimedata.adapter.CharacterAdapter;
import com.example.myanimedata.api.ApiClient;
import com.example.myanimedata.api.ApiService;
import com.example.myanimedata.api.CharacterResponse;
import com.example.myanimedata.api.CharacterResult;
import com.example.myanimedata.databinding.FragmentFavoriteCharacterBinding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class FavoriteCharacterFragment extends Fragment {
    private ApiService apiService;
    private FragmentFavoriteCharacterBinding binding;
    private CharacterAdapter favoriteCharacterAdapter;

    private int page = 1;

    private final List<CharacterResult> favoriteCharacterResults = new ArrayList<>();

    public FavoriteCharacterFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentFavoriteCharacterBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        Retrofit retrofit = ApiClient.getClient();
        apiService = retrofit.create(ApiService.class);

        GridLayoutManager mLayoutManager = new GridLayoutManager(getContext(), 3);

        RecyclerView rvFavoriteCharacterList = root.findViewById(R.id.rvFavoriteCharacterList);
        favoriteCharacterAdapter = new CharacterAdapter(favoriteCharacterResults, getContext());

        rvFavoriteCharacterList.setLayoutManager(mLayoutManager);
        rvFavoriteCharacterList.setAdapter(favoriteCharacterAdapter);
        getFavoriteCharacterData(page);

        rvFavoriteCharacterList.addOnScrollListener(new RecyclerView.OnScrollListener(){
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy){
                super.onScrolled(recyclerView, dx, dy);

                if(!recyclerView.canScrollVertically(1)){
                    page++;
                    getFavoriteCharacterData(page);
                }
            }
        });

        return root;
    }

    private void getFavoriteCharacterData(int page) {
        int limit = 15;
        String orderBy = "favorites";
        String sort = "desc";

        Call<CharacterResponse> call = apiService.getFavoriteCharacter(page, limit, orderBy, sort);
        call.enqueue(new Callback<CharacterResponse>() {
            @SuppressLint("NotifyDataSetChanged")
            @Override
            public void onResponse(@NonNull Call<CharacterResponse> call, @NonNull Response<CharacterResponse> response) {
                if(response.body() != null){
                    int oldCount = favoriteCharacterResults.size();
                    binding.loadingFavoriteCharacter.setVisibility(View.GONE);
                    binding.rvFavoriteCharacterList.setVisibility(View.VISIBLE);

                    favoriteCharacterResults.addAll(response.body().getCharacterResults());
                    favoriteCharacterAdapter.notifyItemRangeInserted(oldCount, favoriteCharacterResults.size());
                } else if(favoriteCharacterResults.isEmpty()) {
                    binding.loadingFavoriteCharacter.setVisibility(View.GONE);
                    binding.textNoFavoriteResult.setVisibility(View.VISIBLE);
                } else {
                    binding.loadingFavoriteCharacter.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<CharacterResponse> call, @NonNull Throwable t) {
                binding.loadingFavoriteCharacter.setVisibility(View.GONE);
                Toast.makeText(getContext(), t.getMessage() + " cause : "
                        + t.getCause(), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/myanimedata/ui/FavoriteCharacterFragment.kt
package com.example.myanimedata.ui

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myanimedata.R
import com.example.myanimedata.adapter.CharacterAdapter
import com.example.myanimedata.api.ApiClient.client
import com.example.myanimedata.api.ApiService
import com.example.myanimedata.api.CharacterResponse
import com.example.myanimedata.api.CharacterResult
import com.example.myanimedata.databinding.FragmentFavoriteCharacterBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class FavoriteCharacterFragment : Fragment() {
    private var apiService: ApiService? = null
    private var binding: FragmentFavoriteCharacterBinding? = null
    private var favoriteCharacterAdapter: CharacterAdapter? = null
    private var page = 1
    private val favoriteCharacterResults: MutableList<CharacterResult> = ArrayList()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = FragmentFavoriteCharacterBinding.inflate(inflater, container, false)

        val root: View = binding!!.root
        val retrofit = client
        apiService = retrofit!!.create(ApiService::class.java)

        val mPortraitLayoutManager = GridLayoutManager(context, 3)
        val mLandscapeLayoutManager = GridLayoutManager(context, 5)
        val rvFavoriteCharacterList = root.findViewById<RecyclerView>(R.id.rvFavoriteCharacterList)
        favoriteCharacterAdapter = context?.let { CharacterAdapter(favoriteCharacterResults, it) }

        if(resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
            rvFavoriteCharacterList.layoutManager = mLandscapeLayoutManager
        } else {
            rvFavoriteCharacterList.layoutManager = mPortraitLayoutManager
        }

        rvFavoriteCharacterList.adapter = favoriteCharacterAdapter
        rvFavoriteCharacterList.setHasFixedSize(true)
        getFavoriteCharacterData(page)

        rvFavoriteCharacterList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!recyclerView.canScrollVertically(1)) {
                    page++
                    getFavoriteCharacterData(page)
                }
            }
        })
        return root
    }

    private fun getFavoriteCharacterData(page: Int) {
        val limit = 15
        val orderBy = "favorites"
        val sort = "desc"
        val call = apiService!!.getFavoriteCharacter(page, limit, orderBy, sort)

        call!!.enqueue(object : Callback<CharacterResponse?> {
            @SuppressLint("NotifyDataSetChanged")
            override fun onResponse(
                call: Call<CharacterResponse?>,
                response: Response<CharacterResponse?>
            ) {
                if (response.body() != null) {
                    val oldCount = favoriteCharacterResults.size
                    binding!!.loadingFavoriteCharacter.visibility = View.GONE
                    binding!!.rvFavoriteCharacterList.visibility = View.VISIBLE
                    favoriteCharacterResults.addAll(response.body()!!.characterResults)
                    favoriteCharacterAdapter!!.notifyItemRangeInserted(
                        oldCount,
                        favoriteCharacterResults.size
                    )
                } else if (favoriteCharacterResults.isEmpty()) {
                    binding!!.loadingFavoriteCharacter.visibility = View.GONE
                    binding!!.textNoFavoriteResult.visibility = View.VISIBLE
                } else {
                    binding!!.loadingFavoriteCharacter.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<CharacterResponse?>, t: Throwable) {
                binding!!.loadingFavoriteCharacter.visibility = View.GONE
                Toast.makeText(
                    context, t.message + " cause : "
                            + t.cause, Toast.LENGTH_SHORT
                ).show()
            }
        })
    }
}